src/treelearner/serial_tree_learner.cpp

```diff
@@ -775,4 +859,231 @@ void SerialTreeLearner::RecomputeBestSplitForLeaf(int leaf, SplitInfo* split) {
   *split = bests[best_idx];
 }
 
+void SerialTreeLearner::GetLeafMap(Tree* tree) {
+  std::fill(leaf_map_.begin(), leaf_map_.end(), -1);
+  // map data to leaf number
+  const data_size_t* ind = data_partition_->indices();
+#pragma omp parallel for schedule(dynamic)
+  for (int i = 0; i < tree->num_leaves(); ++i) {
+    data_size_t idx = data_partition_->leaf_begin(i);
+    for (int j = 0; j < data_partition_->leaf_count(i); ++j) {
+      leaf_map_[ind[idx + j]] = i;
+    }
+  }
+}
+
+template<bool HAS_NAN>
+void SerialTreeLearner::CalculateLinear(Tree* tree, bool is_refit, const score_t* gradients, const score_t* hessians, bool is_first_tree) {
+  tree->SetLinear(true);
+  int num_leaves = tree->num_leaves();
+  int num_threads = OMP_NUM_THREADS();
+  if (is_first_tree) {
+    for (int leaf_num = 0; leaf_num < num_leaves; ++leaf_num) {
+      tree->SetLeafConst(leaf_num, tree->LeafOutput(leaf_num));
+    }
+    return;
+  }
+
+  // calculate coefficients using the additive method described in https://arxiv.org/pdf/1802.05640.pdf
+  // the coefficients vector is given by
+  // - (X_T * H * X + lambda) ^ (-1) * (X_T * H * y + g_T X)
```

Comment:
thank you for your great work. this line is Eq3 or Eq4 in the paper. I can't understand the `y`.

Reply:
It seems that the code below does not use the 'y' stuff. Since the linear models are fitted after the tree is grown, no half-additive is needed.

Reply:
I only find leave linear model to fit. Is there calc split gain use Eq4. If not use Eq4 (but use lgb default) split gain, the tree split is not corresponded to the loss.

Reply:
Sorry, this comment was for an older version of the code. I've updated it now. It's referring to Eq 3 in the paper.